### ISSUE TITLE
Failed to execute 'removeChild' on 'Node': The node to be removed is not a child of this node.

### ISSUE THREAD

**Issue by pizn** [facebook/draft-js#1320]
The current behavior:

![draftjs](https://user-images.githubusercontent.com/986480/28953130-ad26a74c-7908-11e7-8bdd-4757b6017290.gif)

---

macOS + Chrome 60

---

The same issues with #1089, but nobody care.

**Comment by mmmoussa** [comment]
This looks like a duplicate of #1301

**Comment by flarnie** [comment]
Hi - just to get more info, what IME are you using? Is it an app or something that is part of MacOS? If so what version of macOS are you using?

**Comment by pizn** [comment]
@mmmoussa the IME question it's the big question, I found it let me have a headache 😭 
@flarnie  macOS 10.12.5 and pinyin IME.

---

I try to debugger the problem. I found that something wrong with composition events. 

The onCompositionStart event will get the selection's text data, when keydown, the keydown event data will replace selection's data -- At this time, the DOM is removed. And then onCompositionEnd event will set current IME data into the target DOM -- At this time, it will throw the exception.

**Comment by pizn** [comment]
@mmmoussa @flarnie I need your help 🆘

**Comment by gjulivan** [comment]
has this been solved yet?
i have similar issue with chinese character on draft js,
trying to create a custom hashtag decorator.

but error only occurs on chrome.

**Comment by octatone** [comment]
Been getting reports of this specific "Failed to execute 'removeChild' on 'Node'" again specific to Mac + latest Chrome (69).

**Comment by roundrobin** [comment]
@octatone Did you gathered more insights into this? I'm experiencing the same issue with my draft.js editor.

**Comment by roundrobin** [comment]
I found the problem. The issue is related to an exception that's causing react to blow off and is not shown in the stack trace. In my case, it was related to font awesome (the js part of it) messing with the DOM and react not being able to reconcile properly.

I switched to an icon font that doesn't need JS at all. It's anyway a silly idea, I should have known better.

Hope this helps.

**Comment by margox** [comment]
may this helps:
  ```js
  getSelectedBlocks = () => {
  
    const selectionState =this.state.editorState.getSelection()
    const contentState = this.state.editorState.getCurrentContent()
  
    const startKey = selectionState.getStartKey()
    const endKey = selectionState.getEndKey()
    const isSameBlock = startKey === endKey
    const startingBlock = contentState.getBlockForKey(startKey)
    const selectedBlocks = [startingBlock]
  
    if (!isSameBlock) {
      let blockKey = startKey
  
      while (blockKey !== endKey) {
        const nextBlock = contentState.getBlockAfter(blockKey)
        selectedBlocks.push(nextBlock)
        blockKey = nextBlock.getKey()
      }
    }
  
    return selectedBlocks
  
  }
  
  handleCompositionStart = () => {
  
    if (this.getSelectedBlocks().length > 1) {

      const nextEditorState = EditorState.push(
        this.state.editorState,
        Modifier.removeRange(editorState.getCurrentContent(), editorState.getSelection(), 'backward'),
        'remove-range'
      )

      this.setState({
        editorState: nextEditorState
      })

    }
  
  }
  ```
  
  ```jsx
  <div onCompositionStart={this.handleCompositionStart}><Editor {/*....*/}/></div>
  ```

**Comment by Peyton-lee** [comment]
try to catch
`  componentDidCatch(error, info) {
        // You can also log the error to an error reporting service
        // logErrorToMyService(error, info);
        console.log('componentDidCatch:', this.getContent());
        this.setState({
            editorState: EditorState.createEmpty(draftDecorator)
        });
    }`

**Comment by Peyton-lee** [comment]
try to catch
    componentDidCatch(error, info) {
        // You can also log the error to an error reporting service
        // logErrorToMyService(error, info);
        console.log('componentDidCatch:', this.getContent());
        this.setState({
            editorState: EditorState.createEmpty(draftDecorator)
        });
    }`

**Comment by pizn** [comment]
See this workaround. https://github.com/facebook/react/issues/11538#issuecomment-417504600

**Comment by catouse** [comment]
I found a simple solution,  just add `componentDidCatch` method to your react component class:
```js
class MyComponent extends Component {
    componentDidCatch() {
        this.forceUpdate();
    }

    render() {
        return <Editor .../>
    }
}
```

**Comment by ArkadiyLis** [comment]
Hi everyone! On my project I completely get rid from next error "Failed to execute 'removeChild' on 'Node' with next solution
I hope it will help anyone who will face with the same problem.

```
import React from 'react';
import {Editor, EditorState, Modifier} from 'draft-js';

const HANDLED = 'handled';

class AwesomeFixEditor extends React.Component {

    state = {
        editorState: EditorState.createEmpty()
    };

    handleOnChange = editorState => this.setState({editorState});

    handleBeforeInput = (chars, editorState, handlerFunctions) => {
        const currentContentState = handlerFunctions.getEditorState().getCurrentContent();
        const selectionState = handlerFunctions.getEditorState().getSelection();

        this.handleOnChange(EditorState.push(
            editorState,
            Modifier.replaceText(
                currentContentState,
                selectionState,
                chars
            )
        ));

        return HANDLED;
    };

    render() {
        const {editorState} = this.state;

        return (
            <Editor
                editorState={editorState}
                handleBeforeInput={this.handleBeforeInput}
                onChange={this.handleOnChange} />
        );
    }
}
```

**Comment by hjpunzalan** [comment]
> Hi everyone! On my project I completely get rid from next error "Failed to execute 'removeChild' on 'Node' with next solution
> I hope it will help anyone who will face with the same problem.
> 
> ```
> import React from 'react';
> import {Editor, EditorState, Modifier} from 'draft-js';
> 
> const HANDLED = 'handled';
> 
> class AwesomeFixEditor extends React.Component {
> 
>     state = {
>         editorState: EditorState.createEmpty()
>     };
> 
>     handleOnChange = editorState => this.setState({editorState});
> 
>     handleBeforeInput = (chars, editorState) => {
>         const currentContentState = editorState.getCurrentContent();
>         const selectionState = editorState.getSelection();
> 
>         this.handleOnChange(EditorState.push(
>             editorState,
>             Modifier.replaceText(
>                 currentContentState,
>                 selectionState,
>                 chars
>             )
>         ));
> 
>         return HANDLED;
>     };
> 
>     render() {
>         const {editorState} = this.state;
> 
>         return (
>             <Editor
>                 editorState={editorState}
>                 handleBeforeInput={this.handleBeforeInput}
>                 onChange={this.handleOnChange} />
>         );
>     }
> }
> ```

Using removes my toggled inline styles... how can i fix this?

**Comment by jkhaui** [comment]
@hjpunzalan I haven't tried this particular block of code, but `replaceText` accepts an optional 4th argument where you can put the inline style(s) to be retained: https://draftjs.org/docs/api-reference-modifier#replacetext
I think the styles have to be passed as an Immutable Map

**Comment by PulkitKoolKanya** [comment]
@hjpunzalan Your solution is not working with mention plugin. Actually, when I start typing my suggestion for "@" and then backspace the suggested value, the screen will become blank for mobile-only but working on a laptop(Failed to execute 'removeChild' on 'Node': The node to be removed is not a child of this node.). Can you suggest how we can use resolve this?

**Comment by albertodarblade** [comment]
I have the same issue @hjpunzalan  @PulkitKoolKanya  could you help me please

**Comment by bitbasher-dev** [comment]
Someone knows the solution?

**Comment by icellan** [comment]
Bump. Anyone?

**Comment by mifist** [comment]
@icellan @albertodarblade @hjpunzalan @PulkitKoolKanya  Has anyone managed to solve the problem? I have the same error now

**Comment by icellan** [comment]
@mifist No. We removed js-draft from our projects and replaced it with https://github.com/outline/rich-markdown-editor
Users are very happy with the replacement.

**Comment by mifist** [comment]
> ```js
> class MyComponent extends Component {
>     componentDidCatch() {
>         this.forceUpdate();
>     }
> 
>     render() {
>         return <Editor .../>
>     }
> }
> ```

Yes this solution work for me!
I created me custom MentionEditor and wrapped it on this code.
